File: src/commonMain/kotlin/com/scoreo/application/AddGameTypeUseCase.kt
package com.scoreo.application

import com.scoreo.domain.DomainError
import com.scoreo.domain.model.GameType
import com.scoreo.domain.model.WinCondition
import com.scoreo.domain.port.GameTypeRepository

private const val MAX_NAME_LENGTH = 50

class AddGameTypeUseCase(private val repository: GameTypeRepository) {
    operator fun invoke(name: String, winCondition: WinCondition): GameType {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) throw DomainError.Validation("name", "Game type name must not be blank")
        require(trimmed.length <= MAX_NAME_LENGTH) { "Game type name must be $MAX_NAME_LENGTH characters or less" }
        val gameType = GameType(id = IdGenerator.newId(), name = trimmed, winCondition = winCondition)
        repository.save(gameType)
        return gameType
    }
}

File: src/application/addGameTypeUseCase.ts
import { ValidationError } from '../domain/model/errors'
import type { TieBreakRule, WinCondition } from '../domain/model/enums'
import type { GameType } from '../domain/model/gameType'
import type { GameTypeRepository } from '../domain/port/gameTypeRepository'
import { newId } from './idGenerator'

const MAX_NAME_LENGTH = 50

export interface AddGameTypeOptions {
  tieBreakRule?: TieBreakRule
  tieBreakCondition?: WinCondition
  tieBreakLabel?: string | null
}

export class AddGameTypeUseCase {
  constructor(private readonly repository: GameTypeRepository) {}

  invoke(name: string, winCondition: WinCondition, options: AddGameTypeOptions = {}): GameType {
    const trimmed = name.trim()
    if (trimmed.length === 0) {
      throw new ValidationError('name', 'Game type name must not be blank')
    }
    if (trimmed.length > MAX_NAME_LENGTH) {
      throw new ValidationError('name', `Game type name must be ${MAX_NAME_LENGTH} characters or less`)
    }
    const gameType: GameType = {
      id: newId(),
      name: trimmed,
      winCondition,
      tieBreakRule: options.tieBreakRule ?? 'NONE',
      tieBreakCondition: options.tieBreakCondition ?? 'HIGHEST_SCORE',
      tieBreakLabel: options.tieBreakLabel ?? null,
      active: true,
    }
    this.repository.save(gameType)
    return gameType
  }
}
